Valid `.rpm` file: `error reading RPM payload: xz: data is corrupt`

Please review the [Community Note](https://github.com/trufflesecurity/trufflehog/blob/main/.github/community_note.md) before submitting

### TruffleHog Version
3.78.0

### Trace Output
```
2024-06-05T23:13:18-04:00       error   trufflehog      error processing RPM files      {"source_manager_worker_id": "Pe0Ht", "repo": "https://github.com/openshift/origin-web-console-smoke-test.git", "commit": "a6a22f8", "path": "google-chrome-stable-59.rpm", "timeout": 30, "error": "error reading RPM payload: xz: data is corrupt"}
```

https://github.com/trufflesecurity/trufflehog/blob/467c4232c9644e5895d6042ad88caef777d46356/pkg/handlers/rpm.go#L81

### Expected Behavior

The RPM file should be extracted and scanned.

### Actual Behavior

The RPM file is not extracted due to an error.

```sh
$ rpm2cpio google-chrome-stable-59.rpm  | cpio -idmv
./etc/cron.daily/google-chrome
./opt/google/chrome
./opt/google/chrome/chrome
...
./usr/share/man/man1/google-chrome.1
466184 blocks
```

### Steps to Reproduce

<!--- Please list the steps required to reproduce the issue. --->
 1. Download https://github.com/openshift/origin-web-console-smoke-test/blob/a6a22f8/google-chrome-stable-59.rpm
 2. Run trufflehog
 3. Observe reported error

## Environment
 N/A

## Additional Context
N/A

### References
N/A

This seems to be the same issue as #2927.

**Reproducer**
```go
func TestRpm(t *testing.T) {
	r := map[string]func() io.Reader{
		"os.Open": func() io.Reader {
			f, err := os.Open("/tmp/google-chrome-stable-59.rpm")
			require.NoError(t, err)
			return f
		},
		"BufferedFileReader": func() io.Reader {
			f, err := os.Open("/tmp/google-chrome-stable-59.rpm")
			require.NoError(t, err)
			f2, rErr := readers.NewBufferedFileReader(f)
			require.NoError(t, rErr)
			return f2
		},
	}

	for name, reader := range r {
		t.Run(name, func(t *testing.T) {
			var rpm *rpmutils.Rpm
			rpm, err := rpmutils.ReadRpm(reader())
			require.NoError(t, err)

			var reader rpmutils.PayloadReader
			reader, err = rpm.PayloadReaderExtended()
			require.NoError(t, err)

			for {
				fileInfo, err := reader.Next()
				if err != nil {
					if errors.Is(err, io.EOF) {
						return
					}
					require.NoError(t, err)
				}

				fileSize := fileInfo.Size()
				fmt.Printf("FileSize: %d\n", fileSize)
			}
		})
	}
}
```

**Logs**
```
=== RUN   TestRpm/os.Open
FileSize: 15142
...
    --- PASS: TestRpm/os.Open (5.31s)
=== RUN   TestRpm/BufferedFileReader
FileSize: 15142
FileSize: 4096
FileSize: 118169624
    main_test.go:66: 
        	Error Trace:	/tmp/trufflehog/main_test.go:66
        	Error:      	Received unexpected error:
        	            	xz: data is corrupt
        	Test:       	TestRpm/BufferedFileReader
    --- FAIL: TestRpm/BufferedFileReader (0.99s)
```

--- a/pkg/writers/buffered_file_writer/bufferedfilewriter.go
+++ b/pkg/writers/buffered_file_writer/bufferedfilewriter.go
@@ -285,7 +285,6 @@ func (w *BufferedFileWriter) CloseForWriting() error {
 			return err
 		}
 	}
-
 	return w.file.Close()
 }
 
@@ -312,7 +311,6 @@ func (w *BufferedFileWriter) ReadSeekCloser() (io.ReadSeekCloser, error) {
 		if err != nil {
 			return nil, err
 		}
-
 		return newAutoDeletingFileReader(file), nil
 	}
 

--- a/pkg/writers/buffered_file_writer/bufferedfilewriter.go
+++ b/pkg/writers/buffered_file_writer/bufferedfilewriter.go
@@ -209,6 +209,14 @@ func (w *BufferedFileWriter) Write(data []byte) (int, error) {
 		}
 	}
 
+	// Write any remaining data in the buffer to the file before writing new data.
+	if w.buf.Len() > 0 {
+		if _, err := w.buf.WriteTo(w.file); err != nil {
+			return 0, fmt.Errorf("error flushing buffer to file: %w", err)
+		}
+		w.buf.Reset()
+	}
+
 	n, err := w.file.Write(data)
 	if err != nil {
 		return n, err
@@ -273,11 +281,11 @@ func (w *BufferedFileWriter) CloseForWriting() error {
 	defer w.bufPool.Put(w.buf)
 
 	if w.buf.Len() > 0 {
-		_, err := w.buf.WriteTo(w.file)
-		if err != nil {
+		if _, err := w.buf.WriteTo(w.file); err != nil {
 			return err
 		}
 	}
+
 	return w.file.Close()
 }
 
@@ -304,6 +312,7 @@ func (w *BufferedFileWriter) ReadSeekCloser() (io.ReadSeekCloser, error) {
 		if err != nil {
 			return nil, err
 		}
+
 		return newAutoDeletingFileReader(file), nil
 	}
 

--- a/pkg/writers/buffered_file_writer/bufferedfilewriter.go
+++ b/pkg/writers/buffered_file_writer/bufferedfilewriter.go
@@ -205,7 +205,6 @@ func (w *BufferedFileWriter) Write(data []byte) (int, error) {
 				}
 				return 0, err
 			}
-			w.buf.Reset()
 		}
 	}
 
@@ -214,7 +213,6 @@ func (w *BufferedFileWriter) Write(data []byte) (int, error) {
 		if _, err := w.buf.WriteTo(w.file); err != nil {
 			return 0, fmt.Errorf("error flushing buffer to file: %w", err)
 		}
-		w.buf.Reset()
 	}
 
 	n, err := w.file.Write(data)

--- a/pkg/writers/buffered_file_writer/bufferedfilewriter_test.go
+++ b/pkg/writers/buffered_file_writer/bufferedfilewriter_test.go
@@ -650,10 +650,18 @@ func TestBufferedFileWriter_ReadFrom(t *testing.T) {
 			expectedSize:   1 << 20,
 		},
 		{
-			name:           "Input greater than threshold",
+			name:           "Input slightly greater than threshold",
 			input:          string(make([]byte, defaultThreshold+1)),
 			expectedOutput: string(make([]byte, defaultThreshold+1)),
 			expectedSize:   defaultThreshold + 1,
+		},
+		// Test to ensure that anytime the buffer exceeds the threshold, the data is written to a file
+		// and the buffer is cleared.
+		{
+			name:           "Input much greater than threshold",
+			input:          string(make([]byte, (2*defaultThreshold)+largeBufferSize+1)),
+			expectedOutput: string(make([]byte, (2*defaultThreshold)+largeBufferSize+1)),
+			expectedSize:   (2 * defaultThreshold) + largeBufferSize + 1,
 		},
 	}
 
@@ -667,6 +675,10 @@ func TestBufferedFileWriter_ReadFrom(t *testing.T) {
 			reader := bytes.NewReader([]byte(tc.input))
 			size, err := writer.ReadFrom(reader)
 			assert.NoError(t, err)
+
+			if writer.buf != nil && writer.file != nil {
+				assert.Len(t, writer.buf.Bytes(), 0)
+			}
 
 			err = writer.CloseForWriting()
 			assert.NoError(t, err)

